### PR TITLE
node config source validations

### PR DESCRIPTION
**Pull request by sbezverk**
Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

See for reference issue: #72169

/sig node
/sig api-machinery
/kind cleanup
/release-note-none

```release-note
None
```

### CODE REVIEW

File: pkg/registry/core/node/strategy.go [hunk 1]
```diff
@@ -68,9 +68,7 @@ func (nodeStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
 	node := obj.(*api.Node)
 	// Nodes allow *all* fields, including status, to be set on create.
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig) {
```
Comment (liggitt) [inline pkg/registry/core/node/strategy.go]: why was the feature gate check removed?

File: pkg/registry/core/node/strategy.go [hunk 2]
```diff
@@ -131,22 +139,30 @@ func (nodeStatusStrategy) PrepareForCreate(ctx context.Context, obj runtime.Obje
...
```
Comment (liggitt) [inline pkg/registry/core/node/strategy.go]: Why was the feature gate check removed?

File: pkg/registry/core/node/strategy.go [hunk 3]
```diff
@@ -131,22 +139,30 @@ func (nodeStatusStrategy) PrepareForCreate(ctx context.Context, obj runtime.Obje
...
 	newNode.Spec = oldNode.Spec
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig) && !nodeStatusConfigInUse(oldNode) {
 		newNode.Status.Config = nil
-		oldNode.Status.Config = nil
 	}
 }
 
+// nodeConfigSourceInUse returns true if node's Spec ConfigSource is set(used)
```
Comment (liggitt) [inline pkg/registry/core/node/strategy.go]: Update comment and logic to reflect status

File: pkg/registry/core/node/strategy.go (hunk 2 above)
Comment (sbezverk) [reply pkg/registry/core/node/strategy.go]: restored

File: pkg/registry/core/node/strategy.go (hunk 1 above)
Comment (sbezverk) [reply pkg/registry/core/node/strategy.go]: Restored, even though I am not sure why to keep it.

File: pkg/registry/core/node/strategy.go (hunk 3 above)
Comment (sbezverk) [reply pkg/registry/core/node/strategy.go]: done

File: pkg/registry/core/node/strategy.go [hunk 4]
```diff
@@ -67,7 +67,6 @@ func (nodeStrategy) AllowCreateOnUpdate() bool {
 func (nodeStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
 	node := obj.(*api.Node)
 	// Nodes allow *all* fields, including status, to be set on create.
-
 	if !utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig) {
 		node.Spec.ConfigSource = nil
```
Comment (liggitt) [inline pkg/registry/core/node/strategy.go]: also set `node.Status.Config = nil` here

node creation allows setting status, which is unusual (and only allowed for backwards compatibility with early resource behavior)

File: pkg/registry/core/node/strategy.go [hunk 5]
```diff
@@ -130,7 +139,6 @@ var StatusStrategy = nodeStatusStrategy{Strategy}
 func (nodeStatusStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
```
Comment (liggitt) [inline pkg/registry/core/node/strategy.go]: remove `nodeStatusStrategy#PrepareForCreate`... the node status API doesn't allow creation and this is confusing

File: pkg/registry/core/node/strategy.go (hunk 4 above)
Comment (sbezverk) [reply pkg/registry/core/node/strategy.go]: done

File: pkg/registry/core/node/strategy.go (hunk 5 above)
Comment (sbezverk) [reply pkg/registry/core/node/strategy.go]: done

**Pull request merged**